FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedresourcename.md

# Allocated Resource Name

The Allocated Resource Name is a display name which cost is being allocated to in a provider-calculated split cost allocation. The Allocated Resource Name is used to understand what the cost is being allocated to in [*charges*](#glossary:charge) where the provider is allocating costs to something other than the charge's [ResourceID](#ResourceId).

The AllocatedResourceName column adheres to the following requirements:

* AllocatedResourceName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.
* AllocatedResourceName MUST be of type String.
* AllocatedResourceName MUST conform to [StringHandling](#stringhandling) requirements.
* AllocatedResourceName nullability is defined as follows:
  * AllocatedResourceName MUST be null when a charge is not related to a provider-calculated split cost allocation.

>   * AllocatedResourceName MUST NOT be null when a charge is related to a provider-calculated split cost allocation.

Reviewer: ```suggestion
* AllocatedResourceName nullability is defined as follows:
  * AllocatedResourceName MUST be null when [AllocatedResourceId](#allocatedresourceid) is null.
  * AllocatedResourceName MUST NOT be null when AllocatedResourceId is not null.
```

In [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files#r2318478128), I suggested that it seems we agreed that for all Allocated-* columns except AllocatedMethodId, the following should apply:

- [AllocatedColumn] MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.

For AllocatedResourceName, we have two options:

A) Link nullability directly to AllocatedResourceId (*see relevant ResourceName requirements below*)
B) Adjust the requirements similarly to AllocatedResourceId

My current suggestion follows the ResourceName requirements (option A) while taking into account that "...AllocatedResourceName MAY duplicate AllocatedResourceId..." and the assumption that we want to retain MUST NOT be null, as originally specified.

---

**Relevantnih ResourceName requirements:**

> * *ResourceName nullability is defined as follows:*
>   * *ResourceName MUST be null when [ResourceId](#resourceid) is null or when the resource does not have an assigned display name.*
>   * *ResourceName MUST NOT be null when ResourceId is not null and the resource has an assigned display name.*
> * *ResourceName MUST NOT duplicate ResourceId when the *resource* is not provisioned interactively or only has a system-generated ResourceId.*

* AllocatedResourceName MAY duplicate AllocatedResourceId when a separate display name is not applicable.

## Column ID

AllocatedResourceName

## Display Name

Allocated Resource Name

## Description

> The Allocated Resource Name is a display name which cost is being allocated to in a provider-calculated split cost allocation.

Reviewer 2: ```suggestion
The display name of the object to which cost is allocated in provider-calculated split cost allocation.
```
